### DISCUSSION TITLE
Passing services into Domain layer.

### DISCUSSION THREAD

**Discussion by ichensky** [kgrzybek/modular-monolith-with-ddd#136] [Q&A] [answered]
How would you implement for ex. `Edit` method if you needed to load some data from an external service through the `IExternalSystem` anti-corruption layer and add logging? 
Would you pass `IExternalSystem`, `ILogger` as parameters into the method or any other ideas?

https://github.com/kgrzybek/modular-monolith-with-ddd/blob/ac359f221c1589a800fc0fe8048e0cbee4c9d67b/src/Modules/Meetings/Domain/MeetingComments/MeetingComment.cs#L74
```csharp
   public void Edit(MemberId editorId, string editedComment, MeetingCommentingConfiguration meetingCommentingConfiguration, IExternalSystem system, ILogger logger)
        {
            this.CheckRule(new CommentTextMustBeProvidedRule(editedComment));
            this.CheckRule(new MeetingCommentCanBeEditedOnlyByAuthorRule(this._authorId, editorId));
            this.CheckRule(new CommentCanBeEditedOnlyIfCommentingForMeetingEnabledRule(meetingCommentingConfiguration));

            _comment = editedComment;
            _editDate = SystemClock.Now;

            // Here is more logic with `_comment` and other Entities
            // ...

           // Loading some data from the external system
            var data = system.ProcessComment(_comment);
            if(data.SomeState)
            {
                   logger.Warring("Some important information");
                   _comment = data.CommentPrefic + _comment + "_xx";
                  // Do some logic with other Entities
                  // ...
            }

            // Here is more logic with `_comment` and other Entities
            // ...

            this.AddDomainEvent(new MeetingCommentEditedDomainEvent(this.Id, editedComment));
        }
```

**Comment by AndreiGanichev** [comment, marked as answer]
Hi!
You should keep your domain model **as isolated as possible**. This means it must operate only primitive types(or domain primitives) and domain objects and have no dependecies to any external logic(call database or 3rd party services). This constraint helps to express properly ubiquotous language, makes domain model easier to understand(not to add accidential complexity to essential one) and test.
_Repository_ and _Domain service_ are part of domain model but it's also recommended not to pass them to aggregates[1]: 

>Use a repository or domain service to look up dependent objects **ahead** of invoking the aggregate behavior. A client application service may control this, then dispatch to the aggregate.

So it's **application layer's concern** to prepare all data which is needed for business decisions in an use case and then delegate this decisions to domain model. So in the example you should get data from `IExternalSystem` in application logic(_CommandHandler_) and then pass all data to domain logic(`Edit `method). But sometimes this approach has performance drawbacks. To learn more about it I strongly recommend you this post [2]. If application logic is not strightforward and use case depends on the answer of external service you can introduce domain service to incapsulate business decision[3]

As for logging it's a **cross-cutting concern** and doesn't belong to domain logic nor application logic. It's often implemented with decorator pattern. For example you could create decorator for `IExternalSystem` and log there it's answer.

1. Vaughn Vernon. Implementing Domain-Driven Design. Chapter 10. Aggregates
2. Vladimir Khorikov. [Domain model purity vs. domain model completeness](https://enterprisecraftsmanship.com/posts/domain-model-purity-completeness/)
3. Vladimir Khorikov. [Domain services vs Application services](https://enterprisecraftsmanship.com/posts/domain-vs-application-services/)

**Reply by kgrzybek** [reply]
Great, great answer @AndreiGanichev - I have nothing to add 💯

**Reply by AndreiGanichev** [reply]
Thanks! :)
